src/indexer/demuxer.rs

```diff
@@ -0,0 +1,343 @@
+};
+
+/// DemuxMapping can be used to reorganize data from multiple segments.
+/// e.g. if you have two tenant ids TENANT_A and TENANT_B and two segments with
+/// the documents (simplified)
+/// Seg 1 [TENANT_A, TENANT_B]
+/// Seg 2 [TENANT_A, TENANT_B]
+///
+/// You may want to group your documents to
+/// Seg 1 [TENANT_A, TENANT_A]
+/// Seg 2 [TENANT_B, TENANT_B]
+///
+/// Demuxing is the tool for that.
+/// Semantically you can define a mapping from [old segment ordinal, old docid] -> [new segment ordinal].
+#[derive(Debug, Default)]
+pub struct DemuxMapping {
+    /// [index old segment ordinal] -> [index docid] = new segment ordinal
+    mapping: Vec<DocidToSegmentOrdinal>,
+}
+
+/// DocidToSegmentOrdinal maps from docid within a segment to the new segment ordinal for demuxing.
+///
+/// For every source segment there is a `DocidToSegmentOrdinal` to distribute its docids.
+#[derive(Debug, Default)]
+pub struct DocidToSegmentOrdinal {
+    docid_index_to_segment_ordinal: Vec<SegmentOrdinal>,
+}
+
+impl DocidToSegmentOrdinal {
+    /// Creates a new DocidToSegmentOrdinal with size of num_docids.
+    /// Initially all docids point to segment ordinal 0 and need to be set
+    /// the via `set` method.
+    pub fn new(num_docids: usize) -> Self {
+        let mut vec = vec![];
+        vec.resize(num_docids, 0);
+
+        DocidToSegmentOrdinal {
+            docid_index_to_segment_ordinal: vec,
+        }
+    }
+
+    /// Associated a docids with a the new SegmentOrdinal.
+    pub fn set(&mut self, docid: u32, segment_ordinal: SegmentOrdinal) {
+        self.docid_index_to_segment_ordinal[docid as usize] = segment_ordinal;
+    }
+
+    /// Iterates over the new SegmentOrdinal in the order of the docid.
+    pub fn iter(&self) -> impl Iterator<Item = &SegmentOrdinal> {
+        self.docid_index_to_segment_ordinal.iter()
+    }
+}
+
+impl DemuxMapping {
+    /// Creates a new empty DemuxMapping.
+    pub fn empty() -> Self {
+        DemuxMapping {
+            mapping: Default::default(),
+        }
+    }
+
+    /// Creates a DemuxMapping from existing mapping data.
+    pub fn new(mapping: Vec<DocidToSegmentOrdinal>) -> Self {
+        DemuxMapping { mapping }
+    }
+
+    /// Adds a DocidToSegmentOrdinal. The order of the pus calls
+    /// defines the old segment ordinal. e.g. first push = ordinal 0.
+    pub fn add(&mut self, segment_mapping: DocidToSegmentOrdinal) {
+        self.mapping.push(segment_mapping);
+    }
+
+    /// Returns the old number of segments.
+    pub fn get_old_num_segments(&self) -> usize {
+        self.mapping.len()
+    }
+}
+
+fn get_delete_bitsets(
+    demux_mapping: &DemuxMapping,
+    target_segment_ordinal: SegmentOrdinal,
+    max_value_per_segment: &[u32],
+) -> Vec<AliveBitSet> {
+    let mut bitsets: Vec<_> = max_value_per_segment
+        .iter()
+        .map(|max_value| BitSet::with_max_value_and_full(*max_value))
```

Comment:
The inverse would make more sense.

```
BitSet::with_max_value
```

and insert matching docs rather than the other way around.

Reply:
true, should be much faster